Update to Rust Edition 2024

sisou opened the pull request:
Requires Rust v1.85.0, so can only be merged when that version is 2 versions old, as per MSRV policy (this PR sets the MSRV in the CI test to 1.85.0).

## What's in this pull request?

- Remove now unnecessary `ref` and `ref mut` keywords in `match` patterns
- Makes sure the new keyword `gen` is properly escaped where unavoidable
- Satisfy new lifetime declaration rules

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

sisou committed:
Update to Rust edition 2024

Requires rust v1.85.0

nibhar commented:
Cool :+1:

sisou commented:
Apparently the formatter has some new rules now, @hrxi. Uppercase is now sorted first, for example.

Should we apply those new formatting rules, or can you configure them to not reorder all the imports?

hrxi commented:
> or can you configure them to not reorder all the imports?

You can [set `style_edition = "2021"` in rustfmt.toml](https://doc.rust-lang.org/edition-guide/rust-2024/rustfmt-style-edition.html), so we can upgrade these as separate steps at our convenience. I can open a PR.

jsdanielh commented on rustfmt.toml:
```diff
@@ -1,2 +1,3 @@
+style_edition = "2021"
```
Already included in #3321.

hrxi replied:
Will automatically go away when rebasing.

jsdanielh commented on rpc-interface/src/types.rs:
```diff
@@ -319,7 +319,7 @@ impl Block {
                 producer: Slot::from_block_hash(blockchain, &block_hash)?,
                 equivocation_proofs,
-                justification: micro_block.justification.map(Into::into),
+                justification: micro_block.justification,
```
Might also get included by #3319.

sisou replied:
#3324

jsdanielh reviewed (commented):
Missing the change of the MSRV check in [`.github/workflows/build%2Btest.yml#L29`](https://github.com/nimiq/core-rs-albatross/blob/albatross/.github/workflows/build%2Btest.yml#L29).

hrxi commented on database/benches/hash_keys.rs:
```diff
@@ -48,13 +48,13 @@ struct Address([u8; 20]);
 impl Distribution<Blake2bHash> for Standard {
     fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Blake2bHash {
-        Blake2bHash(rng.gen())
+        Blake2bHash(rng.r#gen())
```
Probably better to change to rand 0.9, where this is called `rng.random()`.

sisou replied:
I started doing that on branch `rand-0.9` based on `albatross`, but there are some trait changes that I don't understand. If you could look at the compiler errors, that would be great! :pray:

sisou replied:
#3335

sisou committed:
Set MSRV

hrxi commented:
> Requires Rust v1.85.0, so can only be merged when that version is 2 versions old, as per MSRV policy (this PR sets the MSRV in the CI test to 1.85.0).

This will happen on 2025-05-15, when Rust 1.87 becomes stable.

jsdanielh commented:
Rebased branch. Now that stable is 1.87, we can merge this PR

sisou commented:
New WASM incompatibility. Seems similar to what we had before: https://github.com/rustwasm/wasm-pack/issues/1441#issuecomment-2886314925

jsdanielh commented:
> New WASM incompatibility. Seems similar to what we had before: [rustwasm/wasm-pack#1441 (comment)](https://github.com/rustwasm/wasm-pack/issues/1441#issuecomment-2886314925)

Yeah, let me try using the workaround described there

jsdanielh committed:
ci: Use rust toolchain 1.86.0 for wasm

Use rust toolchain 1.86.0 since the latest stable release (1.87.0)
is making `wasm-opt` unable to parse the produced wasm code.

jsdanielh commented:
> > New WASM incompatibility. Seems similar to what we had before: [rustwasm/wasm-pack#1441 (comment)](https://github.com/rustwasm/wasm-pack/issues/1441#issuecomment-2886314925)
> 
> Yeah, let me try using the workaround described there

The workaround described there didn't work. There is probably more incompatibilities than the ones described so I ended up setting the rust toolchain to 1.86.0 in the meantime.
